File: Android/aiconversationkit/src/main/java/com/trtc/uikit/aiconversationkit/manager/net/ClientAIConversationRequest.java
package com.trtc.uikit.aiconversationkit.manager.net;

import android.text.TextUtils;
import android.util.Log;

import com.google.gson.Gson;
import com.tencent.qcloud.tuicore.TUILogin;
import com.trtc.uikit.aiconversationkit.store.AIConversationConfig;
import com.trtc.uikit.aiconversationkit.common.Logger;
import com.trtc.uikit.aiconversationkit.view.conversation.ConversationConstant;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.TimeZone;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;

public class ClientAIConversationRequest implements AIConversationRequest {
    private static final String TAG                          = "ClientAIConversationReq";
    private static final String VERSION                      = "2019-07-22";
    private static final String ACTION_START_AI_CONVERSATION = "StartAIConversation";
    private static final String ACTION_STOP_AI_CONVERSATION  = "StopAIConversation";

    private final OkHttpClient mClient    = new OkHttpClient();
    private final Gson         mGson      = new Gson();
    private       String       mSecretId  = "";
    private       String       mSecretKey = "";
    private       String       mRegion    = "";
    private volatile String    mTaskId    = "";

    @Override
    public void startConversation(String roomId, AIConversationConfig config) {
        mSecretId = config.getSecretId();
        mSecretKey = config.getSecretKey();
        mRegion = config.getRegion();
        String body = generateStartConversationBody(roomId, config);
        Logger.i(TAG, String.format("startConversation roomId:%s userId:%s aiRobotId:%s", roomId,
                TUILogin.getUserId(), config.getAgentConfig().getAiRobotId()));

        Runnable startAIConversationRun = () -> {
            Request request = buildRequest(mSecretId, mSecretKey, VERSION, ACTION_START_AI_CONVERSATION, body,
                    mRegion, "");
            String response = "";
            try {
                ResponseBody responseBody = mClient.newCall(request).execute().body();
                response = responseBody.string();
            } catch (IOException e) {
                Logger.e(TAG, String.format("startConversation response IOException : %s", e.getMessage()));
            } catch (NullPointerException ne) {
                Logger.e(TAG, String.format("startConversation response NullPointerException : %s", ne.getMessage()));
            }
            Logger.i(TAG, String.format("startAIConversation response : %s", response));
            if (TextUtils.isEmpty(response)) {
                return;
            }
            try {
                JSONObject jsonObject = new JSONObject(response);
                mTaskId = jsonObject.getJSONObject("Response").getString("TaskId");
            } catch (JSONException e) {
                Logger.e(TAG, String.format("conversationTaskId JSONException : %s", e.getMessage()));
            }
        };
        new Thread(startAIConversationRun).start();
    }

    @Override
    public void stopConversation() {
        String body = generateStopConversationBody();
        Logger.i(TAG, String.format("stopConversation : %s", body));

        Runnable stopAIConversationRun = () -> {
            Request request = buildRequest(mSecretId, mSecretKey, VERSION, ACTION_STOP_AI_CONVERSATION, body,
                    mRegion, "");
            String response = "";
            try {
                ResponseBody responseBody = mClient.newCall(request).execute().body();
                response = responseBody.string();
            } catch (IOException e) {
                Logger.e(TAG, String.format("stopConversation response IOException : %s", e.getMessage()));
            } catch (NullPointerException ne) {
                Logger.e(TAG, String.format("stopConversation response NullPointerException : %s", ne.getMessage()));
            }
            Logger.i(TAG, String.format("stopConversation response : %S", response));
            mTaskId = "";
        };
        new Thread(stopAIConversationRun).start();
    }

    private String generateStartConversationBody(String roomId, AIConversationConfig config) {
        HashMap<String, Object> agentMap = new HashMap<>();
        agentMap.put("UserId", config.getAgentConfig().getAiRobotId());
        agentMap.put("UserSig", config.getAgentConfig().getAiRobotSig());
        agentMap.put("TargetUserId", TUILogin.getUserId());
        String welcomeMessage = config.getAgentConfig().getWelcomeMessage();
        if (!TextUtils.isEmpty(welcomeMessage)) {
            agentMap.put("WelcomeMessage", welcomeMessage);
        }
        agentMap.put("InterruptMode", config.getAgentConfig().getInterruptMode());
        agentMap.put("SubtitleMode", 1);

        HashMap<String, Object> sttMap = new HashMap<>();
        String asrLanguage = config.getSttConfig().getLanguage();
        if (!TextUtils.isEmpty(asrLanguage)) {
            sttMap.put("Language", asrLanguage);
        }
        sttMap.put("VadLevel", config.getSttConfig().getVadLevel());
        sttMap.put("VadSilenceTime", config.getSttConfig().getVadSilenceTime());

        HashMap<String, Object> dataMap = new HashMap<>();
        dataMap.put("SdkAppId", TUILogin.getSdkAppId());
        dataMap.put("RoomId", roomId);
        dataMap.put("RoomIdType", ConversationConstant.ROOM_TYPE_STRING);
        dataMap.put("AgentConfig", agentMap);
        dataMap.put("STTConfig", sttMap);
        dataMap.put("LLMConfig", config.getLlmConfig());
        dataMap.put("TTSConfig", config.getTtsConfig());
        return mGson.toJson(dataMap);
    }

    private String generateStopConversationBody() {
        HashMap<String, Object> dataMap = new HashMap<>();
        dataMap.put("TaskId", mTaskId);
        return mGson.toJson(dataMap);
    }

    private Request buildRequest(String secretId,
                                 String secretKey,
                                 String version,
                                 String action,
                                 String body,
                                 String region,
                                 String token) {
        String host = "trtc.tencentcloudapi.com";
        String url = "https://" + host;
        String contentType = "application/json; charset=utf-8";
        String timestamp = java.lang.String.valueOf(System.currentTimeMillis() / 1000);
        String auth = "";
        try {
            auth = getAuth(secretId, secretKey, host, contentType, timestamp, body);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            Log.e(TAG, String.format("getAuth Exception : %s", e.getMessage()));
        }
        return new Request.Builder()
                .header("Host", host)
                .header("X-TC-Timestamp", timestamp)
                .header("X-TC-Version", version)
                .header("X-TC-Action", action)
                .header("X-TC-Region", region)
                .header("X-TC-Token", token)
                .header("X-TC-RequestClient", "SDK_JAVA_BAREBONE")
                .header("Authorization", auth)
                .url(url)
                .post(RequestBody.create(body, MediaType.get(contentType)))
                .build();
    }

    private String getAuth(String secretId,
                           String secretKey,
                           String host,
                           String contentType,
                           String timestamp,
                           String body) throws NoSuchAlgorithmException, InvalidKeyException {
        String canonicalUri = "/";
        String canonicalQueryString = "";
        String canonicalHeaders = "content-type:" + contentType + "\nhost:" + host + "\n";
        String signedHeaders = "content-type;host";

        String hashedRequestPayload = sha256Hex(body.getBytes(StandardCharsets.UTF_8));
        String canonicalRequest = "POST" + "\n" + canonicalUri + "\n" + canonicalQueryString + "\n"
                + canonicalHeaders + "\n" + signedHeaders + "\n" + hashedRequestPayload;

        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
        sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
        String date = sdf.format(new Date(Long.valueOf(timestamp + "000")));
        String service = host.split("\\.")[0];
        String credentialScope = date + "/" + service + "/" + "tc3_request";
        String hashedCanonicalRequest = sha256Hex(canonicalRequest.getBytes(StandardCharsets.UTF_8));
        String stringToSign = "TC3-HMAC-SHA256\n" + timestamp + "\n" + credentialScope + "\n" + hashedCanonicalRequest;
        byte[] secretDate = hmac256(("TC3" + secretKey).getBytes(StandardCharsets.UTF_8), date);
        byte[] secretService = hmac256(secretDate, service);
        byte[] secretSigning = hmac256(secretService, "tc3_request");
        String signature = printHexBinary(hmac256(secretSigning, stringToSign)).toLowerCase();
        return "TC3-HMAC-SHA256 " + "Credential=" + secretId + "/" + credentialScope + ", "
                + "SignedHeaders=" + signedHeaders + ", " + "Signature=" + signature;
    }

    private String sha256Hex(byte[] b) throws NoSuchAlgorithmException {
        MessageDigest md;
        md = MessageDigest.getInstance("SHA-256");
        byte[] d = md.digest(b);
        return printHexBinary(d).toLowerCase();
    }

    private byte[] hmac256(byte[] key, String msg) throws NoSuchAlgorithmException, InvalidKeyException {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKeySpec = new SecretKeySpec(key, mac.getAlgorithm());
        mac.init(secretKeySpec);
        return mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
    }

    private String printHexBinary(byte[] data) {
        StringBuilder r = new StringBuilder(data.length * 2);
        char[] hexCode = "0123456789ABCDEF".toCharArray();
        for (byte b : data) {
            r.append(hexCode[(b >> 4) & 0xF]);
            r.append(hexCode[(b & 0xF)]);
        }
        return r.toString();
    }
}

File: iOS/AIConversationKit/Manager/ClientAIConversationRequest.swift
//
//  ClientAIConversationRequest.swift
//  AIConversationKit
//
//  Created by einhorn on 2024/10/22.
//

import Foundation
import TUICore
import CryptoKit

class ClientAIConversationRequest: AIConversationRequest {
    func sha256(msg: String) -> String {
        guard let data = msg.data(using: .utf8) else { return "" }
        let digest = SHA256.hash(data: data)
        return digest.compactMap{String(format: "%02x", $0)}.joined()
    }
    
    override func start(completion: @escaping (_ taskId: String, _ robotId: String?) -> Void) {
        Logger.info("startConversation userId:\(TUILogin.getUserID() ?? "")"
                    + " aiRobotId:\(config?.agentConfig?.aiRobotId ?? "")")
        
        guard let secretId = config?.secretId, !secretId.isEmpty else {
            completion("", nil)
            return
        }
        guard let secretKey = config?.secretKey, !secretKey.isEmpty else {
            completion("", nil)
            return
        }
        let token = ""
        
        let service = "trtc"
        let host = "trtc.tencentcloudapi.com"
        let endpoint = "https://\(host)"
        let action = "StartAIConversation"
        let version = "2019-07-22"
        let algorithm = "TC3-HMAC-SHA256"
        let timestamp = Int(Date().timeIntervalSince1970)
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.timeZone = TimeZone(identifier: "UTC")
        let date = dateFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(timestamp)))
        
        // ************* step1 *************
        let httpRequestMethod = "POST"
        let canonicalUri = "/"
        let canonicalQuerystring = ""
        let ct = "application/json; charset=utf-8"
        guard let payload = aiParamsConvert2Json(config: config) else {
            completion("", nil)
            return
        }
        let canonicalHeaders = "content-type:\(ct)\nhost:\(host)\n"
        let signedHeaders = "content-type;host"
        let hashedRequestPayload = sha256(msg: payload)
        let canonicalRequest = httpRequestMethod + "\n"
        + canonicalUri + "\n"
        + canonicalQuerystring + "\n"
        + canonicalHeaders + "\n"
        + signedHeaders + "\n"
        + hashedRequestPayload
        
        // ************* step2 *************
        let credentialScope = "\(date)/\(service)/tc3_request"
        let hashedCanonicalRequest = sha256(msg: canonicalRequest)
        let stringToSign = [algorithm, String(timestamp), credentialScope, hashedCanonicalRequest].joined(separator: "\n")
        
        // ************* step3 *************
        let keyData = Data("TC3\(secretKey)".utf8)
        let dateData = Data(date.utf8)
        var symmetricKey = SymmetricKey(data: keyData)
        let secretDate = HMAC<SHA256>.authenticationCode(for: dateData, using: symmetricKey)
        
        let serviceData = Data(service.utf8)
        symmetricKey = SymmetricKey(data: Data(secretDate))
        let secretService = HMAC<SHA256>.authenticationCode(for: serviceData, using: symmetricKey)
        
        let signingData = Data("tc3_request".utf8)
        symmetricKey = SymmetricKey(data: secretService)
        let secretSigning = HMAC<SHA256>.authenticationCode(for: signingData, using: symmetricKey)
        
        let stringToSignData = Data(stringToSign.utf8)
        symmetricKey = SymmetricKey(data: secretSigning)
        let signature = HMAC<SHA256>.authenticationCode(for: stringToSignData, using: symmetricKey).map{String(format: "%02hhx", $0)}.joined()
        
        // ************* step4 Authorization *************
        let authorization = "\(algorithm) Credential=\(secretId)/\(credentialScope), SignedHeaders=\(signedHeaders), Signature=\(signature)"
        
        // ************* step5 start *************
        guard let url = URL(string: endpoint) else {
            completion("", nil)
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = Data(payload.utf8)
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue(authorization, forHTTPHeaderField: "Authorization")
        request.setValue(action, forHTTPHeaderField: "X-TC-Action")
        request.setValue(version, forHTTPHeaderField: "X-TC-Version")
        request.setValue(config?.region ?? "ap-beijing", forHTTPHeaderField: "X-TC-Region")
        request.setValue(String(timestamp), forHTTPHeaderField: "X-TC-Timestamp")
        request.setValue(token, forHTTPHeaderField: "X-TC-Token")
        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            Logger.info("startConversation result data:\(String(describing: data)) response:\(String(describing: response)) error:\(String(describing: error))")
            if let error = error {
                Logger.error("startConversation network error:\(error.localizedDescription)")
                completion("", nil)
                return
            }
            if let data = data {
                do {
                    guard let jsonObject = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
                        Logger.info("startConversation parse jsonObject failed")
                        completion("", nil)
                        return
                    }
                    if let response = jsonObject["Response"] as? [String: Any],
                       let taskId = response["TaskId"] as? String {
                        Logger.info("startConversation result taskId:\(taskId)")
                        completion(taskId, nil)
                    } else {
                        Logger.error("startConversation get taskId failed, response:\(String(describing: String(data: data, encoding: .utf8)))")
                        completion("", nil)
                    }
                } catch {
                    Logger.error("startConversation catch error:\(error.localizedDescription)")
                    completion("", nil)
                }
            } else {
                completion("", nil)
            }
        }
        
        task.resume()
    }
    
    override func stop(taskID: String) {
        Logger.info("stopConversation taskId:\(taskID)")
        guard let secretId = config?.secretId, !secretId.isEmpty else { return }
        guard let secretKey = config?.secretKey, !secretKey.isEmpty else { return }
        let token = ""
        
        let service = "trtc"
        let host = "trtc.tencentcloudapi.com"
        let endpoint = "https://\(host)"
        let action = "StopAIConversation"
        let version = "2019-07-22"
        let algorithm = "TC3-HMAC-SHA256"
        let timestamp = Int(Date().timeIntervalSince1970)
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.timeZone = TimeZone(identifier: "UTC")
        let date = dateFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(timestamp)))
        let httpRequestMethod = "POST"
        let canonicalUri = "/"
        let canonicalQuerystring = ""
        let ct = "application/json; charset=utf-8"
        let payload = "{\"TaskId\":\"\(taskID)\"}"
        let canonicalHeaders = "content-type:\(ct)\nhost:\(host)\n"
        let signedHeaders = "content-type;host"
        let hashedRequestPayload = sha256(msg: payload)
        let canonicalRequest = [httpRequestMethod, canonicalUri, canonicalQuerystring, canonicalHeaders, signedHeaders, hashedRequestPayload].joined(separator: "\n")
        let credentialScope = "\(date)/\(service)/tc3_request"
        let hashedCanonicalRequest = sha256(msg: canonicalRequest)
        let stringToSign = [algorithm, String(timestamp), credentialScope, hashedCanonicalRequest].joined(separator: "\n")
        
        // ************* step3 *************
        let keyData = Data("TC3\(secretKey)".utf8)
        let dateData = Data(date.utf8)
        var symmetricKey = SymmetricKey(data: keyData)
        let secretDate = HMAC<SHA256>.authenticationCode(for: dateData, using: symmetricKey)
        
        let serviceData = Data(service.utf8)
        symmetricKey = SymmetricKey(data: Data(secretDate))
        let secretService = HMAC<SHA256>.authenticationCode(for: serviceData, using: symmetricKey)
        
        let signingData = Data("tc3_request".utf8)
        symmetricKey = SymmetricKey(data: secretService)
        let secretSigning = HMAC<SHA256>.authenticationCode(for: signingData, using: symmetricKey)
        
        let stringToSignData = Data(stringToSign.utf8)
        symmetricKey = SymmetricKey(data: secretSigning)
        let signature = HMAC<SHA256>.authenticationCode(for: stringToSignData, using: symmetricKey).map{String(format: "%02hhx", $0)}.joined()
        let authorization = "\(algorithm) Credential=\(secretId)/\(credentialScope), SignedHeaders=\(signedHeaders), Signature=\(signature)"
        guard let url = URL(string: endpoint) else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = Data(payload.utf8)
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue(authorization, forHTTPHeaderField: "Authorization")
        request.setValue(action, forHTTPHeaderField: "X-TC-Action")
        request.setValue(version, forHTTPHeaderField: "X-TC-Version")
        request.setValue(config?.region ?? "ap-beijing", forHTTPHeaderField: "X-TC-Region")
        request.setValue(String(timestamp), forHTTPHeaderField: "X-TC-Timestamp")
        request.setValue(token, forHTTPHeaderField: "X-TC-Token")
        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            Logger.info("stopConversation result data:\(String(describing: data)) response:\(String(describing: response)) error:\(String(describing: error))")
        }
        
        task.resume()
    }
    
    func aiParamsConvert2Json(config: AIConversationConfig?) -> String? {
        guard let config else { return nil }
        let encoder = JSONEncoder()
        encoder.outputFormatting = .sortedKeys
        do {
            let jsonData = try encoder.encode(config)
            guard var jsonDict = try JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { return nil }
            jsonDict["RoomId"] = roomId
            jsonDict["RoomIdType"] = 1
            jsonDict["SdkAppId"] = Int(TUILogin.getSdkAppID())
            if var agentDict = jsonDict["AgentConfig"] as? [String: Any] {
                agentDict["TargetUserId"] = TUILogin.getUserID() ?? ""
                jsonDict["AgentConfig"] = agentDict
            }
            let resultData = try JSONSerialization.data(withJSONObject: jsonDict, options: [.sortedKeys])
            return String(data: resultData, encoding: .utf8)
        } catch {
            Logger.error("aiParamsConvert2Json failed: \(error)")
            return nil
        }
    }
    
}
